File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/cpp/exercises/practice/dnd-character/dnd_character.cpp
#include "dnd_character.h"
#include <random>
#include <algorithm>
#include <vector>
#include <cmath>

namespace dnd_character {

Character generate() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> die(1, 6);

    Character c;
    c.strength = ability();
    c.dexterity = ability();
    c.constitution = ability();
    c.intelligence = ability();
    c.wisdom = ability();
    c.charisma = ability();

    c.hitpoints = 10 + modifier(c.constitution);

    return c;
}

int modifier(int constitution) {
    return static_cast<int>(std::floor((constitution - 10) / 2.0));
}

int ability() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> die(1, 6);
    
    std::vector<int> rolls(4);
    for (auto& roll : rolls) {
        roll = die(gen);
    }
    std::sort(rolls.begin(), rolls.end(), std::greater<>());
    return rolls[0] + rolls[1] + rolls[2];
}

}  // namespace dnd_character

File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/go/exercises/practice/dnd-character/dnd_character.go
package dndcharacter

import (
	"math"
	"math/rand"
	"sort"
	"time"
)

type Character struct {
	Strength     int
	Dexterity    int
	Constitution int
	Intelligence int
	Wisdom       int
	Charisma     int
	Hitpoints    int
}

// Modifier calculates the ability modifier for a given ability score
func Modifier(score int) int {
	return int(math.Floor(float64(score - 10) / 2))
}

// Ability uses randomness to generate the score for an ability
func Ability() int {
	rand.Seed(time.Now().UnixNano())
	dice := make([]int, 4)
	for i := 0; i < 4; i++ {
		dice[i] = rand.Intn(6) + 1 // Generate dice roll between 1-6
	}
	sort.Ints(dice) // Sort ascending
	return dice[1] + dice[2] + dice[3] // Sum highest 3
}

// GenerateCharacter creates a new Character with random scores for abilities
func GenerateCharacter() Character {
	rand.Seed(time.Now().UnixNano())
	con := Ability() // Need to store constitution separately for HP calculation
	return Character{
		Strength:     Ability(),
		Dexterity:    Ability(),
		Constitution: con,
		Intelligence: Ability(),
		Wisdom:       Ability(),
		Charisma:     Ability(),
		Hitpoints:    10 + Modifier(con),
	}
}
